kube-proxy nftables backend

danwinship opened the pull request:
#### What type of PR is this?
/kind feature
/sig network
/priority important-longterm
/triage accepted

#### What this PR does / why we need it:
Initial implementation of the kube-proxy nftables backend

#### Special notes for your reviewer:
The PR has 4 stages:
1. Make a dummy nftables backend. (The first commit, "`Add a dummy nftables kube-proxy backend which is just a copy of iptables`" ~plus the second commit "`Update nftables/proxier_test.go to use k8s.io/utils/ptr`", which comes from #121154 and is needed to make the linter happy with the "new" code~.) This also includes the feature gate.
2. Simplify the "nftables" backend by removing tricky optimizations, iptables-specific hacks, and features we aren't implementing in nftables. (From "`Simplify nftables/proxier.go by removing large-cluster mode`" to "`Drop unit tests of iptables-specific unit test helpers`").
3. Port the remaining iptables logic over to nftables in a straightforward way. (From "`Create an nftables.Interface in nftables proxier`" to "`Port NAT rules to nftables (and backend is now functional)`")
4. Start improving things by using new nftables functionality. (From "`Redo --nodeport-addresses handling with a set`" to "`Redo service dispatch with maps`")

(We can potentially squash some or all of parts 1-3 if we don't think it's useful to keep that history.)

To do before merging:
- [x] Make it actually work. (An earlier version did, so the general shape is correct, and high-level-reviewable.)
- [x] Clean up iptables rules when starting nftables and nftables rules when starting iptables/ipvs (if feature gate is enabled)
- [x] Ensure it works with the versions of nft in the official k8s builds (Debian stable) and RHEL. Add any necessary fallbacks/workarounds. ([[1]](https://github.com/danwinship/knftables/commit/a17bc5a8565064e44c21582459ac33f0903f62d0), [[2]](https://github.com/danwinship/knftables/commit/32da55dfbfa448b9949dc7f3c9ac89ccda81b7df))


To do before alpha but not necessarily in *this* PR:
- [ ] Fix up metrics
- [ ] Adjust default `--nodeport-addresses` behavior
- [ ] Figure out and document the required kernel/userspace versions
- [ ] Add unit tests for remaining unit test helpers (eg `runPacketFlowTests`) and otherwise ensure iptables-unit-test-parity
- [ ] create an e2e job

To do but not necessarily before alpha:
- [ ] Re-add partial syncing
- [ ] Improved startup error checking relative to iptables
- [ ] perf job
- [ ] Revisit memory allocation perf fixes (eg reusing buffers)
- [ ] Test out `set`-based rather than `mark`-based mark-for-masquerade

#### Does this PR introduce a user-facing change?
(Release note possibly still to be updated if we get good testing data between now and release time...)

```release-note
kube-proxy now has a new nftables-based mode, available by running

    kube-proxy --feature-gates NFTablesProxyMode=true --proxy-mode nftables

This is currently an alpha-level feature and while it probably will not
eat your data, it may nibble at it a bit. (It passes e2e testing but has
not yet seen real-world use.)

At this point it should be functionally mostly identical to the iptables
mode, except that it does not (and will not) support Service NodePorts on
127.0.0.1. (Also note that there are currently no command-line arguments
for the nftables-specific config; you will need to use a config file if
you want to set the equivalent of any of the `--iptables-xxx` options.)

As this code is still very new, it has not been heavily optimized yet;
while it is expected to _eventually_ have better performance than the
iptables backend, very little performance testing has been done so far.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3866
```

danwinship committed:
Add a dummy nftables kube-proxy backend which is just a copy of iptables

danwinship committed:
Simplify nftables/proxier.go by removing Monitor stuff

since it shouldn't be necessary

danwinship committed:
Simplify nftables/proxier.go by removing localhost nodeport support

and related route_localnet setting / anti-martian-packet rule

danwinship committed:
Port service/endpoint chain creation/cleanup to nftables

danwinship committed:
Create an nftables.Interface in nftables proxier

And update most of the comments to refer to "nftables" rather than
"iptables" (even though it doesn't actually do any nftables updating
at this point).

For now the proxy also internally creates a
utiliptablestesting.FakeIPTables to keep the existing sync code
compiling.

danwinship committed:
Replace "iptables-restore" sync in nftables/proxier.go with (trivial) "nft -f -" sync

danwinship committed:
Simplify nftables/proxier.go by removing HaveRandomFully checks

danwinship committed:
Simplify nftables/proxier.go by using string rather than utiliptables.Chain

Change the svcPortInfo and endpointInfo fields to string rather than
utiliptables.Chain, and various fixups from there.

Also use a proper set for activeNATChains, and fix the capitalization
of endpointInfo.chainName.

danwinship committed:
Port filter rules to nftables

danwinship committed:
Port table setup/cleanup code to nftables

danwinship committed:
Clarify nftables/proxier.go by distinguishing nat/filter table KUBE-SERVICES chains

(It is confusing, but allowed, to have distinct "KUBE-SERVICES" chains
in "nat" and "filter" in iptables, but in nftables the "type nat" and
"type filter" chains end up in the same table, so we'll need different
names for the two.)

danwinship committed:
Port NAT rules to nftables (and backend is now functional)

danwinship committed:
Redo LoadBalancerSourceRanges firewall using sets

danwinship committed:
Redo service dispatch with maps

danwinship committed:
Simplify nftables/proxier.go by removing partial syncing

Since optimization will be done differently in nftables.

danwinship committed:
Simplify nftables/proxier.go by removing large-cluster mode

since things will be optimized differently in nftables

danwinship committed:
Simplify nftables/proxier.go by removing the "args" reuse

since that will be done differently in nftables

danwinship committed:
Simplify nftables/proxier.go by dropping "-j ACCEPT" rules

danwinship committed:
Redo no-endpoint handling with maps

danwinship committed:
Drop unit tests of iptables-specific unit test helpers

(We'll eventually have nftables versions.)

danwinship committed:
Redo --nodeport-addresses handling with a set

aojea commented:
one tip @danwinship , meanwhile you are WIP, add a temporal commit that "forces" kube-proxy to run in nftables mode, so it gets exercised in CI and you have quick feedback

aojea commented on pkg/proxy/util/iptables/traffic.go:
```diff
@@ -54,21 +60,39 @@ func (n *noOpLocalDetector) IfNotLocal() []string {
 type detectLocalByCIDR struct {
-	ifLocal    []string
-	ifNotLocal []string
+	ifLocal       []string
+	ifNotLocal    []string
+	ifLocalNFT    []string
+	ifNotLocalNFT []string
```
the interesting value that is the `cidr` or the `interfaceName` is common for both, the other are specific syntaxis details of each implementation, I wanted to change this some time ago to make it more abstract, but I couldn't find anything that improved current state :/

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1542 @@
+)
+
+const (
+	// the nftables table
```
I just noticed we move from uppercase to lowercase on the chains names, is that for some specific reason?

danwinship replied:
iptables and nftables both allow either uppercase or lowercase, but uppercase is the usual convention with iptables and lowercase with nftables... :man_shrugging: 

(We could also do something like "standard chain names are lowercase and per-service/endpoint chain names are uppercase" or vice versa...)

danwinship replied:
> iptables and nftables both allow either uppercase or lowercase, but uppercase is the usual convention with iptables and lowercase with nftables... 🤷‍♂️

or rather: iptables and nftables both allow either uppercase or lowercase letters, and hyphens and underscores, but uppercase-with-hyphens is the usual convention with iptables, and lowercase-with-underscores is the usual convention with nftables.

(And I'm using lowercase-with-hyphens... :scream:.)

thockin replied:
Why lower-with-hyphens if underscores are normal?

danwinship replied:
I don't really have a good answer here. One part of my brain was saying "you should use lowercase because that's normal for nftables" and another part of my brain was saying "you should use hyphens because that's normal for kube-proxy".

I'd be fine changing it, though there's nothing actually _wrong_ with how it is now. It's not going to end up breaking some tools or anything like that, it just doesn't match other people's naming conventions.

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1542 @@
+type servicePortInfo struct {
+	*proxy.BaseServicePortInfo
+	// The following fields are computed and stored for performance reasons.
+	nameString             string
```
you may want to make them this variables typed as the iptables ones. `iptables.Chain` or `iptables.Table` to avoid subtle bugs with the names

danwinship replied:
The table/chain types make sense for APIs like

```
EnsureRule(position RulePosition, table Table, chain Chain, args ...string)
```

where it's useful to syntactically distinguish the table name from the chain name, and both of them from the rule arguments.

But 99% of the time in `pkg/proxy/iptables/proxier.go`, we aren't using those APIs, and we end up doing:

```
				natRules.Write(
					"-A", string(externalTrafficChain),
					"-m", "comment", "--comment", fmt.Sprintf(`"masquerade traffic for %s external destinations"`, svcPortNameString),
					"-j", string(kubeMarkMasqChain))
```

ie, `iptables.Chain` just gets in our way and forces us to typecast it away. And it would be the same in the nftables proxy.

aojea replied:
good point

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1542 @@
+	}
+	return metaproxier.NewMetaProxier(ipv4Proxier, ipv6Proxier), nil
+}
+
```
```suggestion
// Base chains are those that are registered into the 
// https://wiki.nftables.org/wiki-nftables/index.php/Configuring_chains#Adding_base_chains
```

danwinship replied:
ah, yeah, I should probably document better...

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1542 @@
+func servicePortChainNameBase(servicePortName *proxy.ServicePortName, protocol string) string {
+	// nftables chains can contain the characters [A-Za-z0-9_./-]
+
+	// Namespace/Service/Port names are all, roughly /[a-z][a-z-]{0,62}/
```
https://github.com/kubernetes/kubernetes/blob/86a0b1c57a1028729deb82d833007b20955a6cb3/staging/src/k8s.io/apimachinery/pkg/api/validation/generic.go#L60-L63

and 

https://github.com/kubernetes/kubernetes/blob/86a0b1c57a1028729deb82d833007b20955a6cb3/pkg/apis/core/validation/validation.go#L255

This is something we need to be completely sure or this can lead to subtle bugs

danwinship replied:
Oops, I _did_ dig into the validation functions, but then I rewrote the comments a bunch of times (trying to decide how much detail to go into) and ended up with the wrong regexps. It should have said

`// Namespace/Service/Port names are all, roughly /[a-z0-9][a-z0-9-]{0,62}/`

(The difference between `NameIsDNSLabel` and `NameIsDNS1035Label` btw is that the latter requires an initial letter rather than an initial letter-or-number. And neither allows a trailing hyphen, but that doesn't matter for our purposes here.)

I should probably go into more detail in the comments to make it clearer to what extent you could tweak things in the future without breaking things...

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1542 @@
+	servicePortEndpointAffinityNamePrefix   = "affinity-"
+)
+
+// servicePortChainNameBase returns the base name for a chain for the given ServicePort.
```
an example of the output will be nice

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1542 @@
+}
+
+func isServiceChainName(chainString string) bool {
+	// The chains returned from servicePortChainNameBase and
```
are we not prefixing them? `	servicePortPolicyClusterChainNamePrefix = "service-"`

danwinship replied:
We are, but in the middle parts of the PR, there is a chain called `external-services` (corresponding to iptables `KUBE-EXTERNAL-SERVICES`) which would get mistakenly recognized as a service chain because it starts with `external-` (which also corresponds to iptables `KUBE-EXT-`). (The `external-services` chain goes away in a later commit, so I _could_ check prefixes instead after that... Or I could make `external-services` have a different name to avoid the conflict in the first place. But then again, scanning the chain name for a single character seems probably faster than comparing against an array of prefixes?)

Note that the fancy chain names are something I just implemented last week after Tim suggested it in the KEP, so this code is pretty new and maybe less-well-thought-out than some other parts...

aojea commented:
quick look to the entire PR as a whole, not going commit by commit, lots of things to learn from nftables and some comment that may help other readers if you just give the links or expand for references, but it looks pretty neat

The CI fails because `nft` is not installed, at least some of the jobs I checked

```
2023-10-11T01:51:14.247153369Z stderr F E1011 01:51:14.247058       1 proxier.go:1421] "nftables sync failed" err="exec: \"nft\": executable file not found in $PATH"
```

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1542 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
Yeah, I don't know either.

In this case the copyright got copied over because the commit "`Add a dummy nftables kube-proxy backend which is just a copy of iptables`" is literally just

```
mkdir pkg/proxy/nftables
cp pkg/proxy/iptables/proxier*.go pkg/proxy/nftables/
perl -pi -e 's/^package iptables/package nftables/;' pkg/proxy/nftables/*.go
```

There are probably still some references to iptables in nftables/proxier.go at the end of the branch that I have to clean up...

aojea commented:
@danwinship you need to update the existing distroless-iptables image https://github.com/kubernetes/release/blob/7b12f3e7434facb365365cf5843be5dd690401b1/images/build/distroless-iptables/distroless/Dockerfile#L31-L34

danwinship commented:
/hold
for removal of the HACK commit

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1617 @@
+	// right away because if there are map entries pointing to a chain, the chain
+	// doesn't become deletable until a short amount of time *after* the map entry
+	// is deleted. So we flush them now, and record that they need to be deleted
+	// in the next sync.
```
FTR this still isn't quite right; if the BoundedFrequencyRunner allows a "burst" (or the admin sets min-sync-period to 0) then the next sync might be too soon afterward and the deletes will still fail. (This is visible in the latest kind outputs, but I don't think it's why the kind IPv6 job failed.) I'm probably going to have to add an explicit timer or something.

(Also FTR the underlying bug is fixed in very recent kernels though we wouldn't be able to depend on that for a long time...)

danwinship commented:
> pull-kubernetes-unit 	[6723fda](https://github.com/kubernetes/kubernetes/commit/6723fdaae7018ca0f073152a674df8ed4cfba50c) 	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121046/pull-kubernetes-unit/1713962981995319296) 	true 	`/test pull-kubernetes-unit`

Tests that are assuming that iptables is the default mode are being thwarted by the "HACK use nftables mode" commit.

> pull-kubernetes-verify 	[6723fda](https://github.com/kubernetes/kubernetes/commit/6723fdaae7018ca0f073152a674df8ed4cfba50c) 	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121046/pull-kubernetes-verify/1713962984532873216) 	true 	`/test pull-kubernetes-verify`

(gofmt. will be fixed on next push)

> pull-kubernetes-e2e-gce-cos-alpha-features 	[6723fda](https://github.com/kubernetes/kubernetes/commit/6723fdaae7018ca0f073152a674df8ed4cfba50c) 	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121046/pull-kubernetes-e2e-gce-cos-alpha-features/1713962981403922432) 	false 	`/test pull-kubernetes-e2e-gce-cos-alpha-features`

Looks like the "ChromiumOS" kernels don't have nftables support?

```
E1016 17:48:31.917212       7 server.go:547] "Error running ProxyServer" err=<
	unable to create proxier: unable to create ipv4 proxier: could not run nftables command: Error: Could not process rule: Operation not supported
	add table ip kube-proxy
	^^^^^^^^^^^^^^^^^^^^^^^^
 >
```

So we may need to migrate that job to an Ubuntu kernel, or figure out how to reconfigure it appropriately...

aojea commented:
> So we may need to migrate that job to an Ubuntu kernel,

cos jobs are not required, the required one are already using ubuntu

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+	// updating nftables with some partial data after kube-proxy restart.
+	endpointSlicesSynced bool
+	servicesSynced       bool
+	initialized          int32
```
this was a nice addition to the std library https://pkg.go.dev/sync/atomic#Bool, the atomic.bool, it seems you can also remove some mutex with the *synced variables

danwinship replied:
This is just copied from the iptables backend. I do have plans to clean this up eventually: https://github.com/kubernetes/kubernetes/pull/117436/commits/f0e5a48f25eca6506042e0fb228409e349644f86

aojea replied:
great

danwinship replied:
Yes, I think this is another situation where there was initially a single variable that needed to be handled atomically, and then a few more got added and the code as a whole never got reconsidered.

If you look at the commit I linked to above though, my eventual plan is to just get rid of all of the initialization-time synchronization in the Proxiers and just have the Proxier methods not even get called until after the informers are fully synced.

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+func (proxier *Proxier) setupNFTables(tx *knftables.Transaction) {
+	ipX := "ip"
+	ipvX_addr := "ipv4_addr" //nolint:stylecheck // var name intentionally resembles value
+	noLocalhost := "ip daddr != 127.0.0.0/8"
```
is this a subnet? the ipv6 seems to be an ip

danwinship replied:
Yes, this matches the iptables behavior (and general IPv4-vs-IPv6 behavior): IPv6 localhost is just `::1`, but IPv4 localhost is `127.0.0.0/8`.

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+		tx.Flush(&knftables.Set{
+			Name: kubeNodePortIPsSet,
+		})
+		nodeIPs, _ := proxier.nodePortAddresses.GetNodeIPs(proxier.networkInterfacer)
```
we have this in the iptables proxier

```
		nodeIPs, err := proxier.nodePortAddresses.GetNodeIPs(proxier.networkInterfacer)
		if err != nil {
			klog.ErrorS(err, "Failed to get node ip address matching nodeport cidrs, services with nodeport may not work as intended", "CIDRs", proxier.nodePortAddresses)
		}

```

logging helps

danwinship replied:
ah, yes... I think at an earlier point in the branch there are two separate calls to GetNodeIPs per sync, and I log the error from the first but not the second, and when they got merged I must have done it wrong.

aojea commented on vendor/github.com/danwinship/knftables/exec.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+// fakeExec is a mockable implementation of execer for unit tests
+type fakeExec struct {
```
this should go into the exec_test.go , no? so is not build if is not for tests

danwinship replied:
[fixed upstream](https://github.com/danwinship/knftables/commit/3c284e95c5b9ae7ea06c0c3ea7a6de80ea106872), will eventually revendor

aojea commented on vendor/github.com/danwinship/knftables/nftables.go:
```diff
@@ -0,0 +1,444 @@
+	//   ...
+	// ]
+
+	jsonResult := map[string][]map[string]map[string]interface{}{}
```
is this consistent and reliable?

danwinship replied:
what, haven't you ever seen a `map[string][]map[string]map[string]interface{}{}` before??

If you mean, is the JSON syntax consistent and reliable, [yes](https://manpages.debian.org/testing/libnftables1/libnftables-json.5.en.html). Note that we check the `json_schema_version` value too in case that changes.

If you mean, is that code written in the best way it could be, probably not. The JSON schema was unfortunately designed by someone using it from python, so it works well in a dynamically typed language, but terribly in a statically typed language. I would like to find some better way to deal with it...

aojea replied:
> what, haven't you ever seen a `map[string][]map[string]map[string]interface{}{}` before??

I didn't and  I only found one reference here https://grep.app/search?q=map%5Bstring%5D%5B%5Dmap%5Bstring%5Dmap%5Bstring%5Dinterface%7B%7D%7B%7D  and 7 in github https://github.com/search?q=map%5Bstring%5D%5B%5Dmap%5Bstring%5Dmap%5Bstring%5Dinterface%7B%7D%7B%7D&type=code

not my area of knowledge, but is not this something that Unstructured is supposed to solve https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured?

danwinship replied:
> I didn't

That was a rhetorical question. It's obviously terrible. However, it _is_ an accurate representation of the top few levels of the nftables JSON representation.

> is not this something that Unstructured is supposed to solve

No, "Unstructured objects still have functioning TypeMeta features-- kind, version, etc."

aojea replied:
> That was a rhetorical question
lol, got it know

thockin replied:
Barf!

danwinship commented on vendor/github.com/danwinship/knftables/util.go:
```diff
@@ -0,0 +1,117 @@
+// numbers are output as with `fmt.Sprintf("%d")`, and all other types are output as with
+// `fmt.Sprintf("%s")`. To help with set/map lookup syntax, an argument of "@" will not
+// be followed by a space, so you can do, eg, `Concat("ip saddr", "@", setName)`.
+func Concat(args ...interface{}) string {
```
So this is all basically copied from `utilproxy.LineBuffer` of course...

I keep thinking about how we can optimize knftables's memory usage in the face of large transactions, the way that the iptables proxy (a) reuses the same buffers from sync to sync, and (b) reuses the same `[]string` throughout a single sync.

The buffer reusing ought to be easy to add: `realNFTables` could just keep a `bytes.Buffer` around rather than allocating a new one for each transaction. (This needs a little bit of thought to deal with the possibility of multiple pending transactions and stuff, but it's all solvable.)

The equivalent of the `[]string` reusing would be to optimize the way that `Concat()` works...

I realized at some point that if you don't care about reading back the `Rule string` after you construct a `knftables.Rule`, then instead concatenating the rule pieces when you construct the rule, we could just hold off doing it until the transaction gets written out. eg, you'd do something like

```
tx.Add(&knftables.Rule{
        Chain: "mychain",
        Rule: []interface{
                ipX, "daddr", myIP,
                "drop",
        },
})
```

and then the code that's currently in `Concat` would get moved to `Rule.writeOperation`, which would just write them directly into the transaction's `bytes.Buffer` without ever constructing a string containing all of them.

But this is a little uglier in the case where you're constructing a simple `Rule` with a single string... Then I thought maybe generics, and have `type Rule[string | io.Writer] struct { ... Rule T }`, but go doesn't infer generic types in struct constructors the way it does in function calls, so you'd have to explicitly say `&knftables.Rule[string]` or `&knftables.Rule[knftables.ConcatResult]` or whatever.

Maybe we could do

```
        Rule      string
        BuildRule []interface{}
```

and you just set one or the other? and have a `GetRule() string` to simplify the reading-back case...

aojea replied:
is sync.Pool an option https://pkg.go.dev/sync#example-Pool ?

danwinship commented on vendor/github.com/danwinship/knftables/types.go:
```diff
@@ -0,0 +1,338 @@
+
+	// Type is the type of the set key (eg "ipv4_addr"). Either Type or TypeOf, but
+	// not both, must be non-empty.
+	Type string
```
And while I'm pointing out knftables API bits I'm not sure about...

The nftables syntax looks like:

```
add set ip kube-proxy firewall { type ipv4_addr . inet_proto . inet_service }
add element ip kube-proxy firewall { 10.1.2.3 . tcp . 80 }
```

But right now in the knftables API, `Element` has `Key []string`, but `Set` has `Type string`, so you say:

```
tx.Add(&knftables.Set{
        Name: "firewall",
        Type: "ipv4_addr . inet_proto . inet_service",
})
tx.Add(&knftables.Element{
        Set: "firewall",
        Key: []string{"10.1.2.3", "tcp", "80"},
})
```

I thought about changing `Set.Type` to be `[]string` too, for consistency (and so we're not exposing too much nftables syntax in the API), but then this causes a problem with `Map`, because currently `Map.Type` includes both the key type and the value type (e.g. `Type: "ipv4_addr  . inet_proto . inet_service : verdict"`), so we'd have to split that into `KeyType []string, ValueType []string`... But maybe that would be better.

The other tricky part with splitting `Type` is that we have to support `TypeOf` as well (there are certain kinds of values that don't have a named type, so the only way to use them in a set/map key/value is to say "use the type of this expression"). So we'd need `KeyType`, `ValueType`, `KeyTypeOf`, and `ValueTypeOf`, and validate that if you use the type-of variant for one, that you also used it for the other. Or change the way that we handle `type` vs `typeof`, and just have `TypeOf bool` to indicate "use typeof rather than type"?

danwinship replied:
Related: it annoys me that I can do:
```
        Rule: knftables.Concat(
                "ip", "daddr", clusterIP,
                "meta", "l4proto", protocol,
                "th", "dport", svcInfo.Port(),
        ),
```
in a rule, but in an element, I have to do
```
        Key: []string{
                clusterIP.String(),
                protocol,
                strconv.Itoa(svcInfo.Port())
        },
```

aojea replied:
I think that 90% of this feature is the nftables API, the proxier implementation is not something that worries me since is translating the proxier logic to use the nftables "magic" that solves the existing iptables bottleneck, and I think we have a good coverage on testing for it, with the addition that is alpha gated and is really safe to keep iterating.

Is the nft API that worries me, is a direct dependency and is external, I don't trust any userspace library to not have bugs or break compatibility because of reasons, we hit this before https://github.com/kubernetes/kubernetes/issues/112477#issuecomment-1248660064

If this happens and we need to backport any change that means revendoring an entire external project with a new release, with the risks of regressions ... also, since is developed out of tree https://github.com/danwinship/knftables/commits/master we'll not have signal if is compatible until we try to update it.

I think the nft wrapper  should go intree with the proxier ... it can be published externally by using staging/kube-proxy if desired

danwinship replied:
> I think the nft wrapper should go intree with the proxier ... it can be published externally by using staging/kube-proxy if desired

So I know that you and I have had this conversation, but for everyone just joining...

The reason the nftables library is external is that it is _not_ intended to be only/especially for kube-proxy. I already have draft PRs against [CNI](https://github.com/containernetworking/plugins/pull/935) and [ovn-kubernetes](https://github.com/ovn-org/ovn-kubernetes/pull/3709) (which uses iptables for a handful of miscellaneous rules), and plan to get to [cri-o](https://github.com/cri-o/cri-o/blob/v1.28.0/internal/hostport/hostport.go#L135), the non-kube-proxy part of [openshift-sdn](https://github.com/openshift/sdn/blob/release-4.14/pkg/network/node/iptables.go) and a few other more-OpenShift-specific things. We're planning to eventually write a NetworkPolicy implementation using nftables as well. And that's just what _I'm_ working on. My expectation is that this library will be used by other people as well.

> I don't trust any userspace library to not have bugs or break compatibility because of reasons
> ...
> that means revendoring an entire external project with a new release, with the risks of regressions
> ...
> we'll not have signal if is compatible until we try to update it.

These are all arguments for not vendoring any library ever, and that ship has kinda already sailed.

aojea replied:
> So I know that you and I have had this conversation, but for everyone just joining...

yeah, I just thought it will be better if this gets documented on the review 

> The reason the nftables library is external is that it is not intended to be only/especially for kube-proxy. I already have draft PRs against https://github.com/containernetworking/plugins/pull/935 and https://github.com/ovn-org/ovn-kubernetes/pull/3709 (which uses iptables for a handful of miscellaneous rules), and plan to get to [cri-o](https://github.com/cri-o/cri-o/blob/v1.28.0/internal/hostport/hostport.go#L135), the non-kube-proxy part of [openshift-sdn](https://github.com/openshift/sdn/blob/release-4.14/pkg/network/node/iptables.go) and a few other more-OpenShift-specific things.

well, that is something that alleviates my concern, as it will be more users,  the library will benefit of more eyeballs ... for me your argument is fair enough and I'm ok with this approach

danwinship replied:
Oh, the other thing is that I do want to get it out of my `danwinship/` at some point, possibly to `kubernetes-sigs`. (While it's not intended only for kubernetes, it is explicitly intended primarily for kubernetes-adjacent things.) Maybe that should happen at about the point where we bump to `v0.1.0` and a stable-for-now API.

aojea replied:
I'm 100% ok having it in`kubernetes-sigs`, thanks for clarifying

aojea replied:
is reasonable ask for the library to be ~stable as criteria for beta?

danwinship replied:
definitely

thockin replied:
I don't have time today to take apart the whole library, but I thank you for calling out places where it's not up to your bar yet.

aojea commented on cmd/kube-proxy/app/server_others.go:
```diff
@@ -281,6 +282,61 @@ func (s *ProxyServer) createProxier(config *proxyconfigapi.KubeProxyConfiguratio
+		if err != nil {
+			return nil, fmt.Errorf("unable to create proxier: %v", err)
+		}
+	} else if config.Mode == proxyconfigapi.ProxyModeNFTables {
```
has `platformCheckSupported()` checks the iptables requirements and is called from 
https://github.com/kubernetes/kubernetes/blob/369692ab7a037af8adac163246f46a0b1a087e07/cmd/kube-proxy/app/server.go#L641

now it has to depend on the proxyMode , because in a system that uses nftables mode with only nft and not iptables usersapce it will fail, despite it should work.

If you want to fix it later in a follow up then just add a TODO to not forget

danwinship replied:
Added a new commit `Distinguish iptables-based and nftables-based backends, do startup cleanup` which fixes this and also adds the cleanup when switching between iptables-based and nftables-based modes.

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+	ipX := "ip"
+	ipvX_addr := "ipv4_addr" //nolint:stylecheck // var name intentionally resembles value
+	noLocalhost := "ip daddr != 127.0.0.0/8"
+	if proxier.ipFamily == v1.IPv6Protocol {
```
that is unrelated to this PR but to other comment in my PR, v1.IPFamily is an enum with 3 values, and we have "things" that can only be ipv4 or ipv6, I understand that semantically this looks more correct but a boolean is more acccurate, if is not IPv4 then it is IPv6, it can not be unknown

danwinship replied:
`IPFamilyUnknown` is essentially an error value. It should not exist anywhere other than as the return value of a function that is also returning an `error`.

`proxier.ipFamily` _cannot_ be `IPFamilyUnknown`, any more than `proxier.nftables` can be `nil`.

danwinship replied:
(Oh, but also, I don't really love the `ipX`/`ipvX_addr` stuff and am looking for suggestions for improving that... that was a late change after I realized that using `-D IP=ip`/`-D IP=ip6` and `$IP` wouldn't actually work like I thought it would.)

aojea replied:
it seems is a personal taste then

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+// anything from before this transaction.
+func ensureChain(chain string, tx *knftables.Transaction, existing sets.Set[string]) {
+	if existing.Has(chain) {
+		return
```
don't you have to flush it too?

danwinship replied:
No, `existing` is the chains we already created+flushed in this transaction, not the chains that existed before the transaction started. Needs a better name...

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+			Hook:     ptr.To(bc.hook),
+			Priority: ptr.To(bc.priority),
+		}
+		tx.Add(chain)
+		tx.Flush(chain)
```
if the chain already exists , Add() handles the conflict?

aojea replied:
this is also can use `ensureChain`, no?

danwinship replied:
Yes, following the example of the command-line API, `Create` returns `EEXIST` if the object already exists, but `Add` only returns `EEXIST` if an "incompatible" object already exists (e.g., if there's a chain with the same name but a different `Hook`).

(`Delete` returns `ENOENT` if the object you're deleting doesn't exist. In very new versions of the CLI, there is also a `destroy` command that does delete-or-no-op, but I didn't bother trying to emulate that in knftables because there are a few cases where it's complicated. However, you'll see `tx.Add(x); tx.Delete(x)` in a few places, which has the same effect ("ensure that x does not exist") for the types that it works for.)

danwinship replied:
> this is also can use `ensureChain`, no?

no, that doesn't handle base chains

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+			}
+			tx.Add(&knftables.Element{
+				Set: kubeNodePortIPsSet,
+				Key: []string{
+					ip.String(),
+				},
```
is it better to add multiple commands or just pass the array of IPs in one command, or it does not really matter?

danwinship replied:
So this is the API being confusing... the `[]string` here doesn't mean you can pass multiple keys; it means you can pass a key that has multiple components.

suggestions welcome...

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+
+	// Figure out which chains are now stale. Unfortunately, we can't delete them
+	// right away because if there are map entries pointing to a chain, the chain
+	// doesn't become deletable until a short amount of time *after* the map entry is
```
is this an async operation? is there more info on the nature of this "short amount of time"?

danwinship replied:
Oh, I should have mentioned that this was a kernel bug and it's fixed in (very) recent kernels.

I don't have any more details (other than confirmation that, yes, it apparently did work the way I describe, and no, it wasn't intentional). If you look through the older e2e results and corresponding source pushes, you'll see it took me a few tries to fix this. Originally I thought you could do the deletes in a second transaction immediately after the first (ie as part of the same `syncProxyRules`), but that flakes. Then I tried doing them at the start of the _next_ `syncProxyRules`, but that flakes too because we configure the `BoundedFrequencyRunner` to allow "bursts", so you can get two syncs in a row before the rate-limiting kicks in. (I thought about changing that, except that I'm not sure if there's a really good reason why we do it that way, and anyway, an admin might configure a lower `minSyncPeriod` anyway...). So then that led to the current version, where we keep timestamps and only try to delete them after they've been stale for a whole second (which is probably orders of magnitude more time than we actually need to wait).

If we dug into the details of the kernel bug it might turn out that there would be a simpler way to handle the cleanup...

danwinship replied:
Also, note that the timeout turns out to be too short in some circumstances, the result is just that you get warnings in the log and it tries to delete them again later. The stale chains don't actually hurt anything, other than taking up a tiny amount of memory and possibly confusing admins.

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1631 @@
+		}
+	}()
+
+	// If there are sufficiently-stale chains left over from previous transactions,
+	// try to delete them now.
+	if len(proxier.staleChains) > 0 {
+		oneSecondAgo := start.Add(-time.Second)
+		tx := proxier.nftables.NewTransaction()
+		deleted := 0
+		for chain, modtime := range proxier.staleChains {
+			if modtime.Before(oneSecondAgo) {
+				tx.Delete(&knftables.Chain{
+					Name: chain,
+				})
+				delete(proxier.staleChains, chain)
+				deleted++
+			}
+		}
+		if deleted > 0 {
+			klog.InfoS("Deleting stale nftables chains", "numChains", deleted)
+			err := proxier.nftables.Run(context.TODO(), tx)
+			if err != nil {
+				// We already deleted the entries from staleChains, but if
+				// the chains still exist, they'll just get added back
+				// (with a later timestamp) at the end of the sync.
+				klog.ErrorS(err, "Unable to delete stale chains; will retry later")
+				// FIXME: metric
+			}
+		}
+	}
```
this is something unexpected to me, it seems worthy to have a metric to know how many times we hit this problem , of course it can be a follow up

The `1 time.Second` choice has any explanation or just a random value we choose?

aojea replied:
I think this is related to https://github.com/kubernetes/kubernetes/pull/121046#discussion_r1367954679

danwinship committed:
Distinguish iptables-based and nftables-based backends, do startup cleanup

When switching from iptables or ipvs to nftables, clean up old
iptables/ipvs rules. When switching the other way, clean up old
nftables rules.

aojea commented on cmd/kube-proxy/app/server_others.go:
```diff
@@ -100,60 +103,73 @@ func (s *ProxyServer) platformSetup() error {
+	} else {
+		// Assume support for both families.
+		// FIXME: figure out how to check for kernel IPv6 support using nft
+		ipv4Supported, ipv6Supported = true, true
```
don't you need to do at least
```
	nft.path, err = nft.exec.LookPath("nft")
	if err != nil {
		return nil, fmt.Errorf("could not find nftables binary: %w", err)
	}
```

danwinship replied:
(Clarification: in the earliest version of this PR I never actually checked if nft was installed, so if it wasn't, you got a non-obvious error like in https://github.com/kubernetes/kubernetes/pull/121046#issuecomment-1757357455. But I rewrote things so that [`knftables.New()` does sanity checks now](https://github.com/danwinship/knftables/blob/25063956b4f4833424aae4b1020c481876c1e21d/nftables.go#L75-L84), so it will fail at startup now if `nft` is missing or [if it doesn't work for some reason](https://github.com/kubernetes/kubernetes/pull/121046#issuecomment-1765185980).)

aojea replied:
```
I1024 01:39:40.191302       6 server_others.go:306] "Using nftables Proxier"
I1024 01:39:40.191321       6 server_others.go:511] "Detect-local-mode set to NodeCIDR, but no PodCIDR defined at node for family" ipFamily="IPv6"
I1024 01:39:40.191329       6 server_others.go:520] "Defaulting to no-op detect-local"
I1024 01:39:40.191428       6 proxier.go:222] "Using nftables mark for masquerade" ipFamily="IPv4" mark="0x00004000"
E1024 01:39:40.196975       6 server.go:550] "Error running ProxyServer" err=<
	unable to create proxier: unable to create ipv4 proxier: could not run nftables command: Error: Could not process rule: Operation not supported
	add table ip kube-proxy
	^^^^^^^^^^^^^^^^^^^^^^^^
 >
E1024 01:39:40.197008       6 run.go:74] "command failed" err=<
	unable to create proxier: unable to create ipv4 proxier: could not run nftables command: Error: Could not process rule: Operation not supported
	add table ip kube-proxy
	^^^^^^^^^^^^^^^^^^^^^^^^
```

aojea replied:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121046/pull-kubernetes-e2e-gce-cos-alpha-features/1716622550089338880

thockin replied:
is there resolution on this?

danwinship replied:
No, I believe if you try to run kube-proxy in nftables mode on a machine with IPv6 disabled, it will exit with an error (similar to the one Antonio quoted from the ChromiumOS job that can't run nftables at all) rather than falling back back to single-stack IPv4 like it's supposed to.

But as seen above, there's a FIXME in the code about this.

(If you were asking about the ChromiumOS error, that's also not resolved, but it doesn't matter until we make nftables the default mode.)

aojea commented:
/test

aojea commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1623 @@
+	}
+
+	// Now write loadbalancing rule
+	var elements []string
+	for i, ep := range endpoints {
+		epInfo, ok := ep.(*endpointInfo)
+		if !ok {
+			continue
+		}
+
+		elements = append(elements,
+			strconv.Itoa(i), ":", "goto", epInfo.chainName,
+		)
+		if i != len(endpoints)-1 {
+			elements = append(elements, ",")
+		}
+	}
```
this sounds like a kind of helper useful for the library

aojea commented:
/lgtm
/approve
/hold 

you still need to remove d3a1521f114c54f78695333fa95ce62391c8be70

I do not understand nftables enough to review the logic and say, "oh you should use a veredict map here", but the structure is borrowed from the proxier implementation and well documented, the e2e tests are passing and the code is gated in alpha that allow us to iterate and have signal.

IMHO the biggest technical complexity is the nftables library, it has a considerable amount of "machinery". The agreement is for the library to have an stable API for beta https://github.com/kubernetes/kubernetes/pull/121046#discussion_r1368702217

Other topic is where is going to be maintained, I'm personally biased towards having it under the sig-network umbrella in a sig.k8s.io repo per example, but this is very personal opinion, I'm ok with any place that demonstrate the library is well maintained and meet the quality requirements for a component like kube-proxy (having CI, code reviews, ...)

/assign @thockin

danwinship commented:
/hold cancel
(removed the force-nftables hack)

aojea commented:
/lgtm

danwinship commented:
(rebase for `--init-only` and `BaseEndpointInfo` refactor)

danwinship commented:
(rebase for utilptr)

aroradaman commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1633 @@
+	}
+
+	// We currently fully-rebuild our sets and maps on each resync
+	tx.Flush(&knftables.Set{
```
Instead of having `kubeFirewallSet` of type string we can directly define it of type `knftables.Set`, right ?
Then we can use the same variable in every sync rather than creating the structure again to reduce the memory allocations  and deallocations ?

```go
var kubeFirewallSet = &knftables.Set{
    Name: "firewall",
}
```

I see `Set` field of type `knftables.Element` is string type, maybe that should also be of type `&knftables.Set` instead of string (better consistency) ?

Likewise for all the `sets` and `maps`, maybe ?

Just thinking out loud here 🙂

aroradaman commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1633 @@
+	tx.Add(&knftables.Rule{
+		Chain: kubeFirewallCheckChain,
+		Rule: knftables.Concat(
+			ipX, "daddr", ".", "meta l4proto", ".", "th dport", "@", kubeFirewallSet,
```
maybe we can concatenate these strings 
```go
"daddr . meta l4proto . th dport @"
```
instead of
```go
"daddr", ".", "meta l4proto", ".", "th dport", "@"
```
As the buffer is not exposed by library, can't be reset and concatenating will reduce the number of mallocs ... 
https://github.com/danwinship/knftables/blob/fd49603c0a3d6012383657f44cae84781245a1a8/transaction.go#L57

danwinship replied:
I went back and forth on this a little and eventually decided that I liked it better with the arguments split up semantically rather than doing as much "pre-concatenating" as I could.

> As the buffer is not exposed by library, can't be reset(reused) and concatenating will reduce the number of mallocs ...

Of course, we could decide to redo this later if that becomes an issue. See also [this comment](https://github.com/kubernetes/kubernetes/pull/121046#discussion_r1367736334) for other ideas.

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -0,0 +1,1633 @@
+	kubeForwardChain = "forward"
+)
+
+const sysctlNFConntrackTCPBeLiberal = "net/netfilter/nf_conntrack_tcp_be_liberal"
```
Pretty much everything in `pkg/proxy` could benefit from a bit of refactoring...

For now this is just trying to make an nftables backend that looks like the iptables backend and I was mostly trying to avoid additional refactoring in this PR.

thockin commented:
Starting review.  relnote needs love

thockin reviewed (commented):
Dan, your PRs are always a joy to read.  I can see the pleasure in this work, especially the early "delete old crap" commits.

This covers the first 10 commits (up to "Drop unit tests of iptables-specific unit test helpers")

thockin commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 
 //
-// NOTE: this needs to be tested in e2e since it uses iptables for everything.
+// NOTE: this needs to be tested in e2e since it uses nftables for everything.
```
Looking forward - how can we get better and more robust e2e tests of specific k-p functionality.  It feels like that should be seperable from main e2e, no?

thockin commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -231,21 +233,33 @@ func NewProxier(ipFamily v1.IPFamily,
 		endpointsChanges:         proxy.NewEndpointsChangeTracker(hostname, newEndpointInfo, ipFamily, recorder, nil),
 		syncPeriod:               syncPeriod,
-		iptables:                 ipt,
+		iptables:                 utiliptablestesting.NewFake(),
```
Without reading ahead, I can't tell if this is a mistake or a trick or something - why is it fake?

danwinship replied:
comment message explains: it's just to keep everything compiling without having to rip out or comment out all the existing code. It goes away later.

thockin commented on vendor/github.com/danwinship/knftables/README.md:
```diff
@@ -0,0 +1,177 @@
+# knftables: a golang nftables library
```
curious what the "k" stands for?

danwinship replied:
"kool"? :sunglasses: 

It stands for "kubernetes-adjacent".

Originally it was just "nftables" but Antonio pointed out that having a generic name like that can make it hard to refer to, both in code (files that need to import both `"github.com/danwinship/nftables"` and `"k8s.io/kubernetes/pkg/proxy/nftables"` would have to rename one of them) and in conversation (compare https://github.com/cilium/ebpf/discussions/798)

thockin commented on vendor/github.com/danwinship/knftables/nftables.go:
```diff
@@ -0,0 +1,444 @@
+)
+
+// Interface is an interface for running nftables commands against a given family and table.
+type Interface interface {
```
I've soured on this pattern of defining an interface and then returning it.  The Go pattern of "accept interfaces, return types" has started to make more sense to me.

The idea is that the receiver of an argument knows which methods they need, and can/should define a local interface with just those methods.  Trying to encapsulate that here just makes your library more complicated.

Not sure how well it applies here, but "type Interface interface" has become a code-smell to me.

danwinship replied:
Yeah, I'm familiar with "accept interfaces, return types".

The problem here is `knftables.Fake`: we want it to support the same API as the real `knftables.Interface`, and we want consumers to be able to use it as such.

In a very quick search, I couldn't find any "accept interfaces, return types" articles that considered the situation where the API includes a fairly complex mock of itself (which makes heavy use of internal APIs) to help consumers with their own tests.

thockin commented on vendor/github.com/danwinship/knftables/nftables.go:
```diff
@@ -0,0 +1,444 @@
+	if metainfo == nil {
+		return nil, fmt.Errorf("could not find metadata in nft output %q", listOutput)
+	}
+	if version, ok := jsonVal[float64](metainfo, "json_schema_version"); !ok || version != 1.0 {
```
is float64 right or is it going to lose precision?  Should we parse it as string?

danwinship replied:
Is it even possible for us to parse it as anything other than `float64` using `"encoding/json"`? We can't unmarshal the response as a whole into a well-typed struct (or map/array of structs or whatever) like we normally do in k8s, because the response doesn't have consistent enough structure. So as soon as we call `json.Unmarshal` above, the value has been parsed as a `float64`.

But anyway, the schema version is an integer, so we should be good, precision-wise, up until `json_schema_version: 9007199254740992`.

(And as mentioned above in the thread with Antonio, I welcome ideas on how to make this code less awful...)

thockin replied:
A remark that the value is an int would make me happy

thockin reviewed (commented):
nftables library and interface

thockin reviewed (commented):
All the cleanup stuff has historically been a problem - do we really need it?  Are people really live-changing modes without a reboot?

thockin commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -63,29 +63,23 @@ const (
-	// the services chain in the filter table
-	kubeServicesFilterChain = "KUBE-SERVICES"
+	// service dispatch
+	kubeServicesChain  = "services"
```
Should we be prefixing all of these with "kube_" or are they already contextual?  A comment here would help.

danwinship replied:
Everything we do is inside the "kube-proxy" table so the chains/sets don't need any additional prefixing.

thockin reviewed (commented):
more in a bit

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -55,25 +56,25 @@ import (
 
 const (
 	// the services chain
-	kubeServicesChain utiliptables.Chain = "KUBE-SERVICES"
```
IMO the big reason the utiliptables types are good is that without them, we would have APIs like

```
        EnsureRule(string, string, string, ...string)
```

and you'd never be able to remember which argument was which. But having argument-specific types fixes that:

```
        EnsureRule(RulePosition, Table, Chain, ...string)
```

But the knftables APIs are mostly based on structs:

```
        tx.Add(&knftables.Rule{
                Chain:   myChainName,
                Rule:    "my rule ...",
                Comment: ptr.To("my comment"),
        })
````

so they're much more self-documenting / non-confusable even without argument-specific types.

Sure, we _could_ make the `Chain` field there be a `knftables.ChainName` type, but I'm not sure it would help anyone. It would stop people from putting a table name where a chain name belonged, but it wouldn't stop people from putting the _wrong_ chain name there, which seems like a more likely bug...

danwinship commented on go.mod:
```diff
@@ -27,6 +27,7 @@ require (
 	github.com/coreos/go-systemd/v22 v22.5.0
 	github.com/cpuguy83/go-md2man/v2 v2.0.2
 	github.com/cyphar/filepath-securejoin v0.2.4
+	github.com/danwinship/knftables v0.0.12
```
Discussed [here](https://github.com/kubernetes/kubernetes/pull/121046#discussion_r1367748528). I do plan to move it to k-sigs. Antonio and I agreed that should happen before beta.

The reason for not putting it in k/k is that it will be used / is already being used by code outside of k/k. (Some links in the linked comment.)

thockin commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -952,34 +955,34 @@ func (proxier *Proxier) syncProxyRules() {
 				// external traffic may still be accepted.
-				internalTrafficFilterTarget = "DROP"
-				internalTrafficFilterComment = fmt.Sprintf(`"%s has no local endpoints"`, svcPortNameString)
+				internalTrafficFilterVerdict = "drop"
```
given the discussion on this topic, should NFTables mode start with "reject" here?

danwinship replied:
I can change it if you want; other than dropping support for localhost nodeports, I mostly didn't change any of the semantics relative to iptables yet.

thockin replied:
I'm fine with "yet"

thockin commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -1222,36 +1212,67 @@ func (proxier *Proxier) syncProxyRules() {
+					continue
+				}
+
+				// Create a set to store current affinity mappings. The
```
Can we document clearly whether this is interpreting affinity in the iptables way (per-service, per-port) or ipvs way  (per-service) ?

danwinship replied:
Sure. (It's the iptables way.)

thockin reviewed (commented):
affinity commit

danwinship commented:
> All the cleanup stuff has historically been a problem - do we really need it? Are people really live-changing modes without a reboot?

Originally the KEP said that we wouldn't do any automatic cleanup and the admin would be responsible for dealing with it. During PRR review, Maciej noted that "[This seems rather cumbersome](https://github.com/kubernetes/enhancements/pull/3824#discussion_r1326446060)" and so I explained why we do it that way but in the process realized that "why we do it that way" only applies to switching between two different iptables-based modes. (With the old cleanup code, when you started in `iptables` mode, it would call `userspace.CleanupLeftovers`, which would say "oh, `KUBE-SERVICES` is one of my chains, I should delete that!" and end up deleting all of your `iptables`-backend rules.) But when starting up in `nftables` mode, it's definitely safe to delete all iptables rules, and when starting up in `iptables` or `ipvs` mode, it's definitely safe to delete all nftables rules.

(Wojciech was also concerned about making it easy to roll back to iptables if someone tried out nftables and ran into problems, so that also argued for doing the auto-cleanup.)

But anyway, this was a fairly late addition to the branch, and it's also pretty much the only change in this PR that affects the iptables/ipvs codepaths, so if you want to get rid of it (and bring it back in 1.30?) I'm cool with that.

thockin commented:
I skimmed all the commits.  Nothing obviously jumped out as "holy cow, that's terrible", and it's appropriately gated, so I'm ok to say

/approve

What more do we need for merge?

danwinship commented:
> What more do we need for merge?

- `needs-rebase` for `nf_conntrack_tcp_be_liberal`
- a few comments that you requested
- remove the cleanup-on-startup change if you want that removed?
- change the internal traffic policy verdict if you want that changed?

danwinship commented:
First push is just a rebase (there was a conflict between the `utilkernel.GetVersion` changes and the startup-cleanup changes). Second push ([diff](https://github.com/kubernetes/kubernetes/compare/9faff9c6d3015b47fe5e88c31d205ba234d87b77..0993bb78efe4349f5d717e79f204180283746c59)) adds some comments and does a knftables rebase to add a comment there (and belatedly address [this earlier comment](https://github.com/kubernetes/kubernetes/compare/9faff9c6d3015b47fe5e88c31d205ba234d87b77..0993bb78efe4349f5d717e79f204180283746c59), and pick up a bugfix to `knftables.Fake` that doesn't affect us.)

This does not remove the startup-cleanup code...

thockin commented:
I'm game for fixups later.  We can rip out the cleanup if we like, but let's actually discuss it without pressure from a clock.

thockin commented:
/lgtm

danwinship commented:
> relnote needs love

love applied

sftim commented:
Release note suggestion
```markdown
Added a new nftables operating mode for kube-proxy (on Linux).
```

and the extra details would be
```
To use the new kube-proxy nftables mode, start the tool using:
    `kube-proxy --feature-gates NFTablesProxyMode=true --proxy-mode nftables`

This is currently an alpha-level feature and while it probably will not
eat your data, it may nibble at it a bit. (It passes e2e testing but has
not yet seen real-world use.)

At this point it should be functionally mostly identical to the iptables
mode, except that it does not (and will not) support Service NodePorts on
127.0.0.1. (Also note that there are currently no command-line arguments
for the nftables-specific config; you will need to use a configuration file if
you want to set the equivalent of any of the `--iptables-xxx` options.)

As this code is still very new, it has not been heavily optimized yet;
while it is expected to _eventually_ have better performance than the
iptables backend, very little performance testing has been done so far.

To learn 
- 
- [KEP]: https://github.com/kubernetes/enhancements/issues/3866
```

danwinship commented:
> and the extra details would be

what does that mean? Does this go somewhere other than the release note block?

sftim commented:
I meant that for the:
**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
part

&hellip; but I don't actually know how that gets translated into our release notes.

aroradaman commented on vendor/github.com/danwinship/knftables/types.go:
```diff
@@ -0,0 +1,338 @@
+
+	// Comment is an optional comment for the element
+	Comment *string
+}
```
@danwinship 
should we also add a `Elements` type ?
```go
type Elements []Element
```

this would allow to batch all IPs/ports together in one NFT command, reducing total-number of bytes to write and may even improve perf .. as less rules to be parse in total 
```bash
# Element.writeOperation
add element ip kube-proxy firewall { 10.1.2.3 . tcp . 80 }
add element ip kube-proxy firewall { 10.1.2.4 . tcp . 80 }
add element ip kube-proxy firewall { 10.1.2.5 . tcp . 80 }
```

```bash
# Elements.writeOperation
add element ip kube-proxy firewall { 10.1.2.3 . tcp . 80,  10.1.2.4 . tcp . 80, 10.1.2.5 . tcp . 80 }
```

Strictly following the userspace nft API the add element command should accept multiple keys but that won't look neat. 
```go
tx.Add(&knftables.Element{
        Set: "firewall",
        Key: [][]string{{"10.1.2.3", "tcp", "80"},{"10.1.2.4", "tcp", "80"},{"10.1.2.5", "tcp", "80"}},
})
```

Do you think we should try either of the approach to batch the elements of set/map together ?

danwinship replied:
if you want to discuss the knftables API, it's probably better to file issues there?

> this would allow to batch all IPs/ports together in one NFT command, reducing total-number of bytes to write and may even improve perf .. as less rules to be parse in total

eventually we'll be doing partial syncs like in iptables so there'll be a lot less parsing anyway.

Anyway, for most sets, we generally don't know more than 1 (or "a few") elements at a time, because we figure everything out on a service-by-service basis. So we'd have to do a lot of reorganizing to be able to take advantage of something like this.
